src/libguac/protocol.c

```diff
@@ -961,6 +975,50 @@ int guac_protocol_send_rect(guac_socket* socket,
 
 }
 
+/**
+ * Sends the "required" instruction on the given socket, looping
+ * through all the items provided in the NULL-terminated array,
+ * and closing out the instruction.  Returns zero on success, or
+ * non-zero on error.
+ *
+ * @param socket
+ *     The socket on which to write the instruction.
+ *
+ * @param required
+ *     The NULL-terminated array of required parameters to send
+ *     to the client.
+ *
+ * @return
+ *     Zero if successful, non-zero on error.
+ */
+static int __guac_protocol_send_required(guac_socket* socket,
+        const char** required) {
+
+    // The socket should be kept alive while waiting for user response.
+    guac_socket_require_keep_alive(socket);
```

Comment:
So, had a question related to this below on the owner thread, but, question, here - it looks like at least SSH and VNC already set this in the main thread to handle lengthy connects.  Presumably once it's set on a socket it does not have to be re-set - so the usages toward the beginning of the `guac_vnc_client_thread()` and `ssh_client_thread()` functions set it for the duration of the SSH session, correct?  Or do the sockets come and go for a particular connection and it would have to be set, again, during other phases of the connection?

If my assumptions are correct...
* Is there any reason not to just add it to the main RDP thread, prior to authentication?
* With 3 out of 5 protocols implementing the requirement for keep-alives very early on in the connection process, and assuming the keep-alives are then active for the duration of the connections, is there any reason not to just set it across the board?

Reply:
> Presumably once it's set on a socket it does not have to be re-set - so the usages toward the beginning of the `guac_vnc_client_thread()` and `ssh_client_thread()` functions set it for the duration of the SSH session, correct?

Correct.

> Is there any reason not to just add it to the main RDP thread, prior to authentication?

Nope. It imposes very minimal overhead, and is already designed to not bother sending pings unless nothing has been sent along the `guac_socket` in a while.

> With 3 out of 5 protocols implementing the requirement for keep-alives very early on in the connection process, and assuming the keep-alives are then active for the duration of the connections, is there any reason not to just set it across the board?

Actually ... I think this would be fine, too. That fact would need to be documented, but it seems like that would be best. We can then mark `guac_socket_require_keep_alive()` as deprecated and note that it is now a no-op.

Reply:
Is there a central location where the keep alive should be enabled when a socket starts up, or just do it at the beginning of each protocol implementation?

Reply:
I think I got this in the right spot...

Reply:
Hm ... looking over 9bfcbbf3a576054f82ce9bce6ef2a0497304ca8a, I'm not sure `guac_socket` allows for this to be done safely by default:

1. The current architecture of `guac_socket` relies on it being impossible for that `guac_socket` to be used before its handlers have been initialized. This is because a pointer to the new `guac_socket` simply does not exist outside the internals of the various `guac_socket` implementations until they have finished their init process.

   If `guac_socket` will now create this thread as part of its own init (which happens first), then there is no guarantee that the handlers will be properly initialized by the implementation  before the thread starts trying to write `nop` instructions, leading to unstable behavior.

2. Creating the keep-alive ping on the broadcast socket is nice and efficient as it only results in a new single thread, with that thread covering all users of a connection. Creating it across the board would result in a new thread for the broadcast socket plus new threads for each of the per-user sockets it writes to, a bit of a waste.

It would be safe for implementations to start the keep-alive ping on their own, but I don't think `guac_socket` can do this on behalf of all implementations. There's just no way for those implementations to let `guac_socket` know that initialization is complete.

My knee-jerk thoughts on the above would be to say that we should start the keep-alive by default within the broadcast socket implementation and leave it at that.

What do you think?

Reply:
In fact ... given that there are definitely cases where a `guac_socket` does not need a `nop` ping (the tee socket writing screen recordings to a file, for example), perhaps the keep-alive should only be set by default where it is known to be correct, like right after initialization of the `guac_client` via `guac_client_load_plugin()` succeeds within guacd?

Even assuming that the broadcast socket needs a keep-alive ping might be overly presumptuous on the utility of a socket that broadcasts to all users.

Reply:
So, somewhere around here?

https://github.com/apache/guacamole-server/blob/c10ceab7e80f9d6eae9ad98254bf97f41a0de112/src/guacd/proc.c#L316-L329

Is it safe to assume, there, that the socket is correctly initialized and ready to have the `nop` ping started?

Reply:
Went ahead and implemented it, there...let me know if there's somewhere else.  And, thanks, as always, for the patience with me :-).